### ISSUE TITLE
Error opening a TLS connection - Race condition

### ISSUE THREAD

**Issue by yharaskrik** [prisma/prisma#18141]
### Bug description

During some deployment (we have never experienced it locally) we receive the following error

```
Error: Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733: (self signed certificate in certificate chain)
    at startFn (/app/node_modules/@prisma/client/core/runtime/library.js:100:2563)
    at Proxy.onModuleInit (/app/dist/apps/aurora/webpack:/libs/data-access/server/prisma-core/src/lib/base-core-prisma.service.ts:34:9)
    at async Promise.all (index 1)
    at callModuleInitHook (/app/node_modules/@nestjs/core/hooks/on-module-init.hook.js:43:5)
    at NestApplication.callInitHook (/app/node_modules/@nestjs/core/nest-application-context.js:210:13)
    at NestApplication.init (/app/node_modules/@nestjs/core/nest-application.js:97:9)
    at NestApplication.listen (/app/node_modules/@nestjs/core/nest-application.js:155:33)
```

// base-core-prisma.service.ts
```typescript
@Injectable()
export abstract class BaseCorePrismaService<T extends Prisma.PrismaClientOptions = Prisma.PrismaClientOptions>
    extends PrismaClient
    implements OnModuleInit, OnModuleDestroy
{
    protected constructor(
        protected readonly config: BaseCorePrismaConfig,
        protected readonly optionsArg?: Prisma.Subset<T, Prisma.PrismaClientOptions>
    ) {
        super(optionsArg);
    }

    async onModuleInit(): Promise<void> {
        await this.$connect(); // Row 34

        if (!this.config.readReplica) {
            // Enable soft deletes
            this.$use(SoftDeleteMiddleware);
        }
    }

    async onModuleDestroy(): Promise<void> {
        await this.$disconnect();
    }

    async enableShutdownHooks(app: INestApplication): Promise<void> {
        this.$on('beforeExit', async () => {
            await app.close();
        });
    }
}
```

This is an abstract class in our NestJS APIs that our db model wrapper extend so we have shared configuration between the class that connects to our Postgres DB and the class that connects to our MongoDB (thanks to prisma having a unified setup this makes it really easy! thanks!).

This error only happens sometimes (pretty rarely actually, as I thought it was fixed a month ago since we have not seen it since).

Our connection strings are using `&TrustServerCertificate=true&sslmode=require` and the server is deployed in a Docker container based on `FROM node:18.12.1-bullseye` to Elastic beanstalk on t4g.micro instances (Arm-based AWS Graviton2 processors), this does not seem to happen on the x86 elastic beanstalk environments (at least not that we have scene _yet_). the binary targets we have are `binaryTargets = ["native", "linux-musl-openssl-3.0.x", "linux-arm64-openssl-1.1.x", "debian-openssl-1.1.x"]`

### How to reproduce

Reproduction is VERY tricky as it only happens **rarely**, only on ARM instances, and only when deployed to Elastic beanstalk. Open to ideas here I would love to be able to repro it.

### Expected behavior

No issues with self signed certs and connecting.

### Prisma information

```prisma
>3000 lines, if a prisma engineer needs specifics feel free to let me know and I can provide what you need.
```

Removed some stuff for clarity, the Mongo service is using the base service in a similar way to the `CorePrismaService`
```typescript
@Injectable()
export abstract class BaseCorePrismaService<T extends Prisma.PrismaClientOptions = Prisma.PrismaClientOptions>
    extends PrismaClient
    implements OnModuleInit, OnModuleDestroy
{
    protected constructor(
        protected readonly config: BaseCorePrismaConfig,
        protected readonly optionsArg?: Prisma.Subset<T, Prisma.PrismaClientOptions>
    ) {
        super(optionsArg);
    }

    async onModuleInit(): Promise<void> {
        await this.$connect();

        if (!this.config.readReplica) {
            // Enable soft deletes
            this.$use(SoftDeleteMiddleware);
        }
    }

    async onModuleDestroy(): Promise<void> {
        await this.$disconnect();
    }

    async enableShutdownHooks(app: INestApplication): Promise<void> {
        this.$on('beforeExit', async () => {
            await app.close();
        });
    }
}


@Injectable()
export class CorePrismaService extends BaseCorePrismaService {
    constructor(@Inject(CORE_PRISMA_CONFIGURATION) configuration: CorePrismaConfiguration) {
        super(
            {},
            {
                log: configuration.debug ? ['query'] : undefined,
                datasources: {
                    db: configuration.primary,
                },
            }
        );
    }
}
```

### Environment & setup

AWS Linux
t4g.micro Arm-based AWS Graviton2 processors
Docker node:18.12.1-bullseye
AWS RDS Postgres & MongoDB Atlas

### Prisma Version

```
4.10.0
```

**Comment by janpio** [comment]
> as I thought it was fixed a month ago since we have not seen it since

Does that imply the error also happened in older versions before 4.10.0 already? Important for us to pinpoint if this could be something we recently introduced - or not.

**Comment by yharaskrik** [comment]
> > as I thought it was fixed a month ago since we have not seen it since
> 
> Does that imply the error also happened in older versions before 4.10.0 already? Important for us to pinpoint if this could be something we recently introduced - or not.

Yes, it also happened in 4.9.x, I upgraded to 4.10.0 in case that helped.

**Comment by yharaskrik** [comment]
Anything I can provide here to help this? I am still seeing it.

**Comment by janpio** [comment]
Any information that helps us to understand what is happening, optimally a _reliable_ reproduction.

**Comment by yharaskrik** [comment]
> Any information that helps us to understand what is happening, optimally a _reliable_ reproduction.

I will see what I can do.

**Comment by strazi** [comment]
Started seeing exactly this error:
```
Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733
```

On Prisma `4.11.0`, but much of my setup is identical: Docker, ARM, but on Fargate.

I've been on `4.11.0` for about 3 weeks, and this has only happened today. Today I had a fairly high frequency of back to back deploys, so unclear if that's connected.

**Comment by yharaskrik** [comment]
Just tried this on Prisma 4.12.0 which unfortunately had the same issue, I also tried to move the `$connect` to `onApplicationBootstrap` rather than `onModuleInit` in our Nest application which unfortunately didn't seem to help. (the former happens after the latter in the lifecycle)

**Comment by yharaskrik** [comment]
> I found a sentry log today relating to this is seems. One of our Gql data loaders is throwing this error
> 
> ```
> Invalid `prisma.fundraiser.findUnique()` invocation:
> 
> 
> Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733: (self signed certificate in certificate chain)
> ```
> 
> I am wondering if our server (just this one for some reason) is serving requests on the `POST /graphql` endpoint before prisma is fully initialized, we have health checks setup on this server from what I can tell the same way as our other server with GraphQL and Prisma. I will investigate some more.

This might be a red herring, unsure, it seems sentry it not reporting this specifically anymore. Heres the raw stack trace

```
Error: Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733:
  File "/app/node_modules/@prisma/client/core/runtime/library.js", line 101, col 2598, in startFn
    '{snip} =this.parseInitError(i.message);throw typeof o=="string"?i:new G(o.message,this.config.clientVersion,o.error_code)}finally{this.libraryStart {snip}
  File "/app/dist/apps/aurora/webpack:/libs/data-access/server/prisma-core/src/lib/base-core-prisma.service.ts", line 33, col 9, in Proxy.onApplicationBootstrap
  File "/app/node_modules/@prisma/client/core/runtime/library.js", in Promise.all
  File "/app/node_modules/@nestjs/core/hooks/on-app-bootstrap.hook.js", line 43, col 5, in callModuleBootstrapHook
    await Promise.all(callOperator(nonTransientInstances));
  File "/app/node_modules/@nestjs/core/nest-application-context.js", line 230, col 13, in NestApplication.callBootstrapHook
    await (0, hooks_1.callModuleBootstrapHook)(module);
  File "/app/node_modules/@nestjs/core/nest-application.js", line 99, col 9, in NestApplication.init
    await this.callBootstrapHook();
  File "/app/node_modules/@nestjs/core/nest-application.js", line 155, col 33, in NestApplication.listen
    !this.isInitialized && (await this.init());
```

**Comment by nkovacic** [comment]
Getting the same error, also using a `Nest.js` `prisma` wrapper service.
````
          throw new PrismaClientInitializationError(error.message, this.config.clientVersion!, error.error_code)
                ^
Error: Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733:
    at startFn (/workspace/node_modules/@prisma/engine-core/src/library/LibraryEngine.ts:380:17)
    at /workspace/node_modules/@prisma/engine-core/src/tracing/runInChildSpan.ts:41:14
    at Proxy.onModuleInit (/workspace/node_modules/@bird-buddy/nestjs-prisma/src/prisma.service.ts:100:5)
    at async Promise.all (index 0)
    at callModuleInitHook (/workspace/node_modules/@nestjs/core/hooks/on-module-init.hook.js:43:5)
    at NestApplication.callInitHook (/workspace/node_modules/@nestjs/core/nest-application-context.js:227:13)
    at NestApplication.init (/workspace/node_modules/@nestjs/core/nest-application.js:98:9)
    at NestApplication.listen (/workspace/node_modules/@nestjs/core/nest-application.js:168:33)
    at bootstrap (/workspace/dist/webpack:/app-api/src/main.ts:71:3)
````

**Comment by darvelo** [comment]
Also seeing the same error:

```
[user-service] PrismaClientKnownRequestError:
[user-service] Invalid `prisma.user.findMany()` invocation:
[user-service]
[user-service]
[user-service] Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733:, error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733:
[user-service]     at Pn.handleRequestError (/home/node/app/node_modules/@prisma/client/runtime/library.js:171:6929)
[user-service]     at Pn.handleAndLogRequestError (/home/node/app/node_modules/@prisma/client/runtime/library.js:171:6358)
[user-service]     at Pn.request (/home/node/app/node_modules/@prisma/client/runtime/library.js:171:6237)
[user-service]     at async Promise.all (index 5)
[user-service]     at async DataLoader.batchUsersByAccountIds [as _batchLoadFn] (file:///home/node/app/dist/src/datasources/user.js:285:16) {
[user-service]   code: 'P1011',
[user-service]   clientVersion: '4.14.1',
[user-service]   meta: {
[user-service]     message: 'error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733:, error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733:'
[user-service]   }
[user-service] }
```

**Comment by ackinc** [comment]
I encountered the same error today. I can reliably reproduce it on an AWS EC2 t4g.micro instance running AMI ubuntu-focal-20.04-arm64, but the error does not occur on a t2.micro running the same OS.

1. Output of `npx prisma --version`

```
prisma                  : 4.16.2
@prisma/client          : 4.16.2
Current platform        : linux-arm64-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/libquery_engine-linux-arm64-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/migration-engine-linux-arm64-openssl-1.1.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.16.1-1.4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Default Engines Hash    : 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Studio                  : 0.484.0
```

2. The script I'm running:

``` index.js
const { PrismaClient } = require("@prisma/client");

let dbUrl = new URL("postgresql://..."); // connects to RDS Proxy
dbUrl.searchParams.append("connection_limit", 200);
dbUrl.searchParams.append("pool_timeout", 0);

const prisma = new PrismaClient({
  datasources: { db: { url: dbUrl.href } },
});

const nQueriesToRun = Number(process.argv[2]);
for (let i = 0; i < nQueriesToRun; i++) {
  prisma.$executeRaw`SELECT pg_sleep(5)`.then(console.log);
}
```

3. The error occurs when executing ~50 or more queries concurrently

**Comment by hiramhuang** [comment]
I have the same error recently. I’m also running on AWS Arm processors (Fargate Arm) with the same Docker image (node:bullseye).

**Comment by strazi** [comment]
I switched my CI and container setup all over to x86 and haven't seen this issue again.

**Comment by hiramhuang** [comment]
I don't think this is a Prisma bug, because I also received TLS errors from `fetch()` on the same environment.

And this issue does not occur in my Lambda function using the `lambda/nodejs:16-arm64` image (which is based on Amazon Linux 2).

It's possible that this is a Debian or OpenSSL bug.

**Comment by hiramhuang** [comment]
I attempted to resolve this issue by using Amazon Linux 2023 for my ECS Docker (remains on Arm) last week. However, the error persisted. 

The error stopped appearing after I changed my ECS Fargate CPU architecture to x86_64 a few days ago.

**Comment by darvelo** [comment]
@aqrln Could this issue possibly also have the same root cause as #18510? They both seem to be OpenSSL issues on arm64.

**Comment by aqrln** [comment]
@darvelo that sounds very plausible, I'm 90% sure it indeed has the same root cause. It might be challenging to try to reproduce this exact error on my side so it would be very helpful if someone who encountered it could test `5.3.0-integration-arm-openssl-5-2-0.1` and see if it fixes the issue.

**Comment by darvelo** [comment]
@aqrln After a couple of days I haven't seen the OpenSSL error happen again, but I did see an entirely new error:

```
PrismaClientKnownRequestError:
Invalid `prisma.project.findMany()` invocation:


Can't reach database server at `database-service-postgres-cluster-rw`:`5432`

Please make sure your database server is running at `database-service-postgres-cluster-rw`:`5432`.
    at vn.handleRequestError (/home/node/app/node_modules/@prisma/client/runtime/library.js:123:6730)
    at vn.handleAndLogRequestError (/home/node/app/node_modules/@prisma/client/runtime/library.js:123:6119)
    at vn.request (/home/node/app/node_modules/@prisma/client/runtime/library.js:123:5839)
    at async l (/home/node/app/node_modules/@prisma/client/runtime/library.js:128:9763)
    at async Promise.all (index 3) {
  code: 'P1001',
  clientVersion: '5.3.0-integration-arm-openssl-5-2-0.1',
  meta: {
    database_host: 'database-service-postgres-cluster-rw',
    database_port: 5432
  }
}
```

Not sure if it's related but it is suspicious that once the OpenSSL bug seems fixed this issue started to occur. FWIW I do a lot of nested includes which may be causing some sort of timeout in Prisma, triggering this error.

**Comment by aqrln** [comment]
@darvelo interesting, did it happen only once or multiple times?

**Comment by darvelo** [comment]
@aqrln Hard to say, the logs don't have timestamps, but it seems it may have happened several times in succession and then stopped happening at all so far for the past day.

**Comment by GeertWille** [comment]
We're seeing the same issues on Fargate, ARM using Prisma 4.16.2.

```
PrismaClientKnownRequestError 
Invalid `prisma.marketplace.findMany()` invocation:


Error opening a TLS connection: error:24067044:random number generator:rand_pool_add:internal error:crypto/rand/rand_lib.c:733: 
    node_modules/.store/@prisma-client-virtual-528490a155/node_modules/@prisma/client/runtime/library.js:174:7325 Rn.handleRequestError
    node_modules/.store/@prisma-client-virtual-528490a155/node_modules/@prisma/client/runtime/library.js:174:6754 Rn.handleAndLogRequestError
    node_modules/.store/@prisma-client-virtual-528490a155/node_modules/@prisma/client/runtime/library.js:174:6344 Rn.request
    index.js:16773:20 async corsMiddleware
```

and also

```
PrismaClientKnownRequestError 
Invalid `prisma.marketplace.findMany()` invocation:


Can't reach database server at `platform-api-production-cluster.cluster-cloetgx9dx8v.eu-central-1.rds.amazonaws.com`:`5432`

Please make sure your database server is running at `xxx.cluster-cloetgxxxx.eu-central-1.rds.amazonaws.com`:`5432`. 
    node_modules/.store/@prisma-client-virtual-528490a155/node_modules/@prisma/client/runtime/library.js:174:7325 Rn.handleRequestError
    node_modules/.store/@prisma-client-virtual-528490a155/node_modules/@prisma/client/runtime/library.js:174:6754 Rn.handleAndLogRequestError
    node_modules/.store/@prisma-client-virtual-528490a155/node_modules/@prisma/client/runtime/library.js:174:6344 Rn.request
    index.js:16773:20 async corsMiddleware
```
 
Can anyone confirm this goes away by upgrading? I really want to stick with ARM but if this is an ongoing problem I might switch back.

**Comment by darvelo** [comment]
@GeertWille I can confirm I haven't seen this issue on ARM (M1) since I upgraded to Prisma 5.3.1.

**Comment by mvpasarel** [comment]
Doesnt work for me in AWS Fargate task. It cant connect to the RDS

```
/app # DEBUG="*" node query.js
  prisma:tryLoadEnv  Environment variables not found at null +0ms
  prisma:tryLoadEnv  Environment variables not found at undefined +10ms
  prisma:tryLoadEnv  No Environment variables loaded +0ms
  prisma:client  checkPlatformCaching:postinstall false +0ms
  prisma:client  checkPlatformCaching:ciName undefined +2ms
  prisma:tryLoadEnv  Environment variables not found at null +11ms
  prisma:tryLoadEnv  Environment variables not found at undefined +1ms
  prisma:tryLoadEnv  No Environment variables loaded +0ms
  prisma:client  dirname /app/node_modules/.prisma/client +0ms
  prisma:client  relativePath ../../../libs/api/data-access-db/src/lib +2ms
  prisma:client  cwd /app/libs/api/data-access-db/src/lib +1ms
  prisma:client  clientVersion 5.3.1 +2ms
  prisma:client:libraryEngine  internalSetup +0ms
  prisma:client  prisma.$queryRaw(SELECT * FROM Company, []) +0ms
  prisma:client  Prisma Client call: +45ms
  prisma:client  prisma.$queryRaw({
  query: "SELECT * FROM Company",
  parameters: {
    values: "[]",
    __prismaRawParameters__: true
  }
}) +6ms
  prisma:client  Generated request: +0ms
  prisma:client  {
  "action": "queryRaw",
  "query": {
    "arguments": {
      "query": "SELECT * FROM Company",
      "parameters": "[]"
    },
    "selection": {}
  }
}
 +2ms
  prisma:client:libraryEngine  sending request, this.libraryStarted: false +53ms
  prisma:get-platform  Found distro info:
{
  "targetDistro": "musl",
  "familyDistro": "alpine",
  "originalDistro": "alpine"
} +0ms
  prisma:get-platform  Don't know any platform-specific paths for "alpine" on x64 (x86_64) +2ms
  prisma:get-platform  Falling back to "ldconfig" and other generic paths +1ms
  prisma:get-platform  Found libssl.so file using "ldconfig" or other generic paths: libssl.so.1.1 +86ms
  prisma:get-platform  The parsed libssl version is: 1.1.x +3ms
  prisma:client:engines:resolveEnginePath  enginePath /app/node_modules/.prisma/client/libquery_engine-linux-musl.so.node +0ms
  prisma:client:libraryEngine  library starting +236ms
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault
```
Prisma 5.3.1

```
/app # prisma version
prisma                  : 5.3.1
@prisma/client          : 5.3.1
Current platform        : linux-musl
Query Engine (Node-API) : libquery-engine 61e140623197a131c2a6189271ffee05a7aa9a59 (at node_modules/@prisma/engines/libquery_engine-linux-musl.so.node)
Schema Engine           : schema-engine-cli 61e140623197a131c2a6189271ffee05a7aa9a59 (at node_modules/@prisma/engines/schema-engine-linux-musl)
Schema Wasm             : @prisma/prisma-schema-wasm 5.3.1-2.61e140623197a131c2a6189271ffee05a7aa9a59
Default Engines Hash    : 61e140623197a131c2a6189271ffee05a7aa9a59
Studio                  : 0.494.0
```

Dockerfile

```
FROM --platform=linux/amd64 node:18.10.0-alpine as base

WORKDIR /app

# set for base and all layer that inherit from it
ENV NODE_ENV production

RUN apk add --no-cache python3 make g++ libc6-compat jq openssl


............


COPY docker-entrypoint.sh /usr/local/bin/
RUN chmod +x /usr/local/bin/docker-entrypoint.sh
ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]

EXPOSE 3333

CMD [ "yarn", "start:api:prod" ]

```

**Comment by janpio** [comment]
@mvpasarel I think you are the first person reporting a segmentation fault in this issue, so I suspect your problem is different from the others and you should open a new issue.

**Comment by aqrln** [comment]
@mvpasarel your issue is a completely different one: you don't get the same error others are reporting, and you are on x86_64 while this issue is explicitly about ARM. Did you open a new issue as @janpio asked you by any chance?

**Comment by aqrln** [comment]
Since we didn't get any new reports, and there are two confirmations above about the issue not occurring anymore since 5.3.x, I'm gonna make a call and say it was fixed by https://github.com/prisma/engine-images/pull/75.
The "Can't reach database server" error that was mentioned in some of the messages above seems unrelated and orthogonal to the error in the original report — please open a new issue if you are still affected by it.
